### ISSUE TITLE
v7+HMMA: GPU dispatch produces stale logits — set_mnist_input value not reflected

### ISSUE THREAD

**Issue by johnjezl** [SLM-OS/SLM-Operating-System#715]
## Summary

After #710 closed (PR #694), v7+HMMA dispatch end-to-end works on jetson-nano-2 — no fastpath_failed log, GPU-derived logits returned to the caller. But: **two different inputs produce identical GPU logits**.

## Reproducer

```bash
# kexec with HMMA + --no-gpu-suspend already done
gpu use inference on

# Input A: synthetic vertical bar (rows 6-21, cols 13-14 = 1.0)
lua-admin /mnt/files/all_logits.lua
# argmax=1, logit[0]=-0.9124786, logit[1]=13.0703344, ...

# Input B: all zeros (784 floats of 0.0)
lua-admin /mnt/files/zero_logits.lua
# argmax=1, logit[0]=-0.9124786, logit[1]=13.0703344, ...   ← SAME
```

Both lua scripts call `slm.model_infer_bytes(0, bytes)` with different `bytes` payloads. The path is:

```
slm.model_infer_bytes → rust_infer → engine::run_inference
  → run_mnist_gpu_fastpath
    → kernel_ffi::gpu_set_mnist_input(input_bytes)   ← receives different bytes
    → kernel_ffi::gpu_run_mnist(&mut logits)
```

`gpu_set_mnist_input` calls `slm_gpu_set_input(KIND_MNIST, ...)` → `ga10b_bringup_set_input` which `memcpy`s into `g_handoff.input_buf_phys` and `cache_clean`s. So the CPU side IS writing the user's bytes to DRAM.

The GPU dispatch reads input via op[0]'s cbuf — which was populated by the helper at handoff time and references some buffer. Suspect: cbuf points at a buffer DIFFERENT from `input_buf_phys`, or the cache_clean isn't reaching the GPU's view through L2.

## Hypotheses

1. **cbuf vs input_buf mismatch.** The helper allocates an input buffer for op[0]'s cbuf to reference. SLM-OS's `set_input` writes to `g_handoff.input_buf_phys` — which may be a separate allocation from the cbuf-referenced buffer. Audit: dump op[0]'s cbuf contents and compare against `input_buf_phys`.

2. **GPU L2 caches the input.** Even with CPU-side `cache_clean` + DSB, the GPU's L2 might hold the old input. The QMD has `INVALIDATE_SHADER_DATA_CACHE_BIT=1` and `CWD_MEMBAR_TYPE=L1_SYSMEMBAR` set — should be enough — but worth verifying empirically by adding a per-dispatch L2_FLUSH_DIRTY (already used in inherit).

3. **The v7 op[0] reads from a per-op input_phys field that the helper baked at handoff time.** v7's `ga10b_pipeline_op_v7` doesn't carry an `input_phys` field, but the helper might construct cbuf with a fixed VA pointing at its own staged input — which `set_input` doesn't update.

## What's NOT broken

- v7+HMMA dispatch path itself — confirmed by the BULK trace and non-CPU-fallback logits
- HMMA SASS execution — confirmed by standalone helper run (max|err|=0.000410 vs CPU)
- Channel inherit, validation, find-by-kind, scanner — all working

## Acceptance criteria

`slm.model_infer_bytes(0, bytes)` produces logits whose values change with the `bytes` payload. Verified by running two distinct inputs (e.g. all-zero and a vertical bar) and observing different logit vectors.

## Related

- PR #694 — landed v7+HMMA dispatch fixes
- #710 — closed: dispatch-side issue (op-1 payload didn't land)
- #702 — `compute_ready` audit (orthogonal)

**Comment by johnjezl** [comment]
**False alarm.** Re-tested on jetson-nano-2 with fresh kexec + fresh uploads + diagnostic instrumentation in `ga10b_bringup_set_input`:

```
[DIAG-715] set_input dst=0x139dcc000 cap=3136 off724=0000803f   # vertical-bar input (1.0 at row 6 col 13)
argmax=1
  logit[0] = -0.9124786, logit[1] = 13.0703344, ...

[DIAG-715] set_input dst=0x139dcc000 cap=3136 off724=00000000   # all-zero input
argmax=5
  logit[0] = -0.0448560, logit[1] = 0.0077917, ..., logit[5] = 0.1402187, ...
```

The diag confirms (a) FFI propagates fresh bytes per call (off724 changes from 0x3f800000=1.0 to 0x00000000), (b) GPU produces input-sensitive logits (vertical-bar predicts '1', all-zero produces tiny logits with argmax=5).

The original observation in the issue ("identical logits across two different inputs") was caused by stale state in the previous test: `slm-put.py` resumed a partial upload, leaving the lua file with mixed old+new content. That file's lua code was a near-no-op that didn't exercise the path I thought it did, so both 'inputs' ended up running effectively-identical inference.

End-to-end v7+HMMA verification on jetson-nano-2 is complete. PR #694 fully unlocks the path:
- Inherit + v7 channel handoff: working
- v7 BULK dispatch (8 ops + sema, 0xcafedead release): working
- HMMA tensor-core GEMM at op 6: working (helper standalone showed max|err|=0.000410 vs CPU)
- Runtime input swap via slm_gpu_set_mnist_input: working
- Different inputs → different logits: working

Closing as not-a-bug. The DIAG-715 instrumentation was stripped before this comment.

**Comment by johnjezl** [comment]
Re-test on jetson-nano-2 with **production kernel from main** (post-merge, no DIAG instrumentation) shows the off-by-one staleness is REAL. My earlier closure used vertical-bar/zero ABBA inputs which can't distinguish off-by-one from correct (the alternating pattern hides the lag).

## Reproducer (real MNIST test digits, model from helper standalone)

```lua
-- ABBA: same two inputs alternating
infer(0, '/mnt/files/digit_0_idx3.bin')
infer(0, '/mnt/files/digit_3_idx18.bin')
infer(0, '/mnt/files/digit_0_idx3.bin')
infer(0, '/mnt/files/digit_3_idx18.bin')
-- AAAA: same input 4 times
infer(0, '/mnt/files/digit_7_idx0.bin')
infer(0, '/mnt/files/digit_7_idx0.bin')
infer(0, '/mnt/files/digit_7_idx0.bin')
infer(0, '/mnt/files/digit_7_idx0.bin')
```

## Output

```
ABBA:
  digit_0 -> argmax=7  (== prior test's last input)
  digit_3 -> argmax=0  (digit_0's result, lagged)
  digit_0 -> argmax=3  (digit_3's result, lagged)
  digit_3 -> argmax=0  (digit_0's result, lagged)
AAAA:
  digit_7 -> argmax=3  (digit_3's lagged)
  digit_7 -> argmax=7  ← correct after 1-iter warmup
  digit_7 -> argmax=7  ← stable
  digit_7 -> argmax=7  ← stable
```

**Pattern: every call returns the result of the PREVIOUS call's input. Same input N+ times → result stabilizes after one lag.**

Same symptom as #558. PR #694's QMD-pool fix forces SKED to redecode the descriptor per launch (so the dispatch *fires*), but the GPU's L2 cache is still returning stale input bytes from the previous dispatch.

## What's NOT the cause

- **Not the FFI**: `gpu_set_mnist_input` is being called with the right bytes — verified by DIAG instrumentation in the prior round (off724 fingerprint changed across calls).
- **Not the QMD encoder**: per-dispatch QMD is fresh; CWD_MEMBAR_TYPE_L1_SYSMEMBAR + INVALIDATE_SHADER_DATA_CACHE_BIT + INVALIDATE_TEXTURE_DATA_CACHE_BIT are all set.
- **Not v7 vs v6 dispatch**: would expect both paths to show the same staleness if root cause is L2.

## Suspected cause

The GPU's chip-wide L2 (LTC) caches input data read via global memory. The QMD's invalidate bits cover SM-level caches (L1, shader data, instruction, constant, texture) but not L2. `CWD_MEMBAR_TYPE_L1_SYSMEMBAR` is supposed to issue a system-memory barrier before kernel start that forces loads through to sysmem — empirically not enough on this hardware.

NVK and nouveau emit a dedicated **MEM_BARRIER** method between consecutive compute dispatches in the pushbuffer (Option 2 from the original analysis at `docs/gpu-qmd-per-dispatch-plan.md` §7). Likely the next fix to try.

## Apologies

Closing prematurely was a process failure. The verification matrix needed at least 3 distinct inputs and a same-input-twice test to surface this. Updating the SLM-OS GPU verification checklist accordingly.
